File: bootstrap/src/main/java/net/dontdrinkandroot/wicket/bootstrap/component/item/DropDownItem.java
/*
 * Copyright (C) 2012-2016 Philip Washington Sorst <philip@sorst.net>
 * and individual contributors as indicated
 * by the @authors tag.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.dontdrinkandroot.wicket.bootstrap.component.item;

import net.dontdrinkandroot.wicket.behavior.CssClassAppender;
import net.dontdrinkandroot.wicket.bootstrap.behavior.DropDownToggleBehavior;
import net.dontdrinkandroot.wicket.bootstrap.behavior.IconBehavior;
import net.dontdrinkandroot.wicket.bootstrap.component.dropdown.DropDownMenu;
import net.dontdrinkandroot.wicket.bootstrap.css.BootstrapCssClass;
import org.apache.wicket.markup.html.basic.Label;
import org.apache.wicket.model.IModel;

/**
 * @author Philip Washington Sorst <philip@sorst.net>
 */
public abstract class DropDownItem<T> extends AbstractLabeledItem<T>
{
    private IconBehavior iconBehavior = new IconBehavior().setAppendIcon(BootstrapCssClass.CARET);

    protected DropDownMenu dropDownMenu;

    public DropDownItem(String id, IModel<String> labelModel)
    {
        this(id, labelModel, null);
    }

    public DropDownItem(String id, IModel<String> labelModel, IModel<T> model)
    {
        super(id, labelModel, model);
    }

    @Override
    protected void onInitialize()
    {
        super.onInitialize();

        this.add(new CssClassAppender(BootstrapCssClass.DROPDOWN));

        Label dropDownToggle = new Label("dropDownToggle", this.getLabel());
        dropDownToggle.add(new DropDownToggleBehavior());
        dropDownToggle.add(this.iconBehavior);
        this.add(dropDownToggle);

        this.dropDownMenu = this.createDropDownMenu("dropDownMenu");
        this.add(this.dropDownMenu);
    }

    public IconBehavior getIconBehavior()
    {
        return this.iconBehavior;
    }

    protected abstract DropDownMenu createDropDownMenu(String id);
}

File: bootstrap/src/main/kotlin/net/dontdrinkandroot/wicket/bootstrap/component/item/DropdownItem.kt
package net.dontdrinkandroot.wicket.bootstrap.component.item

import net.dontdrinkandroot.wicket.behavior.CssClassAppender
import net.dontdrinkandroot.wicket.bootstrap.behavior.DropdownToggleBehavior
import net.dontdrinkandroot.wicket.bootstrap.component.dropdown.DropdownMenu
import net.dontdrinkandroot.wicket.bootstrap.css.BootstrapCssClass
import net.dontdrinkandroot.wicket.bootstrap.css.DropdownAlignment
import net.dontdrinkandroot.wicket.kmodel.model
import org.apache.wicket.behavior.Behavior
import org.apache.wicket.markup.html.link.AbstractLink
import org.apache.wicket.model.IModel
import org.apache.wicket.model.Model

abstract class DropdownItem<T>(
    id: String,
    model: IModel<T>? = null,
    label: IModel<String>,
    vararg linkBehaviors: Behavior,
) : AbstractLinkItem<T, AbstractLink>(id, model, label, emptyArray(), *linkBehaviors) {

    protected lateinit var dropdownMenu: DropdownMenu

    override fun onInitialize() {
        super.onInitialize()
        this.add(CssClassAppender(BootstrapCssClass.DROPDOWN))
        dropdownMenu = createDropdownMenu("dropdownMenu")
        this.add(dropdownMenu)
    }

    override fun createLink(id: String): AbstractLink {
        val link: AbstractLink = object : AbstractLink(id) {}
        link.add(DropdownToggleBehavior())
        return link
    }

    protected abstract fun createDropdownMenu(id: String): DropdownMenu
}

inline fun ItemView.addDropdown(
    label: String,
    dropdownAlignment: DropdownAlignment? = null,
    vararg linkBehaviors: Behavior,
    crossinline populateItemsHandler: ItemView.() -> Any?
): RepeatingDropdownItem<Void> {
    val repeatingDropdownItem = object :
        RepeatingDropdownItem<Void>(this.newChildId(), null, model(label), Model(dropdownAlignment), *linkBehaviors) {
        override fun populateItems(itemView: ItemView) {
            populateItemsHandler(itemView)
        }
    }
    this.add(repeatingDropdownItem)
    return repeatingDropdownItem
}

inline fun ItemView.addDropdown(
    label: String,
    vararg linkBehaviors: Behavior,
    crossinline populateItemsHandler: ItemView.() -> Any?
): RepeatingDropdownItem<Void> {
    val repeatingDropdownItem = object :
        RepeatingDropdownItem<Void>(this.newChildId(), null, model(label), Model(null), *linkBehaviors) {
        override fun populateItems(itemView: ItemView) {
            populateItemsHandler(itemView)
        }
    }
    this.add(repeatingDropdownItem)
    return repeatingDropdownItem
}

inline fun ItemView.addDropdown(
    label: IModel<String>,
    dropdownAlignment: DropdownAlignment? = null,
    vararg linkBehaviors: Behavior,
    crossinline populateItemsHandler: ItemView.() -> Any?
): RepeatingDropdownItem<Void> {
    val repeatingDropdownItem = object :
        RepeatingDropdownItem<Void>(this.newChildId(), null, label, Model(dropdownAlignment), *linkBehaviors) {
        override fun populateItems(itemView: ItemView) {
            populateItemsHandler(itemView)
        }
    }
    this.add(repeatingDropdownItem)
    return repeatingDropdownItem
}